drivers/input/inport/inport.c

```diff
@@ -0,0 +1,239 @@
+        Status = STATUS_NO_SUCH_DEVICE;
+        goto Failure;
+    }
+
+    IoInitializeRemoveLock(&FdoExtension->RemoveLock, INPORT_TAG, 0, 0);
+    IoInitializeDpcRequest(Fdo, InPortDpcForIsr);
+
+    Status = IoGetDeviceProperty(PhysicalDeviceObject,
+                                 DevicePropertyHardwareID,
+                                 sizeof(HardwareIdBuffer),
+                                 HardwareIdBuffer,
+                                 &DummyResultLength);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to query the hardware ID string 0x%X\n", Status);
+        goto Failure;
+    }
+    RtlInitUnicodeString(&HardwareId, HardwareIdBuffer);
+
+    if (RtlEqualUnicodeString(&HardwareId, &HardwareId1, FALSE))
+    {
+        FdoExtension->MouseType = NecBusMouse;
+        FdoExtension->MouseAttributes.SampleRate = 60;
+        FdoExtension->MouseAttributes.NumberOfButtons = 2;
+    }
+    else if (RtlEqualUnicodeString(&HardwareId, &HardwareId3, FALSE) ||
+             RtlEqualUnicodeString(&HardwareId, &HardwareId4, FALSE))
+    {
+        FdoExtension->MouseType = MsInPortMouse;
+        FdoExtension->MouseAttributes.SampleRate = 100;
+        FdoExtension->MouseAttributes.NumberOfButtons = 3;
+    }
+    else if (RtlEqualUnicodeString(&HardwareId, &HardwareId2, FALSE) ||
+             RtlEqualUnicodeString(&HardwareId, &HardwareId5, FALSE) ||
+             RtlEqualUnicodeString(&HardwareId, &HardwareId6, FALSE))
+    {
+        FdoExtension->MouseType = LogitechBusMouse;
+        FdoExtension->MouseAttributes.SampleRate = 100;
+        FdoExtension->MouseAttributes.NumberOfButtons = 3;
+    }
+    else
+    {
+        DPRINT1("Unrecognized hardware '%wZ'\n", &HardwareId);
+        Status = STATUS_DEVICE_REMOVED;
+        goto Failure;
+    }
+    FdoExtension->MouseAttributes.MouseIdentifier = MOUSE_INPORT_HARDWARE;
+    /* 1 packet */
+    FdoExtension->MouseAttributes.InputDataQueueLength = sizeof(MOUSE_INPUT_DATA);
+
+    Fdo->Flags |= DO_BUFFERED_IO;
+    if (FdoExtension->Ldo->Flags & DO_POWER_PAGABLE)
+        Fdo->Flags |= DO_POWER_PAGABLE;
+    Fdo->Flags &= ~DO_DEVICE_INITIALIZING;
+
+    return STATUS_SUCCESS;
+
+Failure:
+    if (FdoExtension)
+    {
+        if (FdoExtension->Ldo)
+            IoDetachDevice(FdoExtension->Ldo);
+    }
+
+    if (Fdo)
+    {
+        IoDeleteDevice(Fdo);
+    }
+
+    return Status;
+}
+
+VOID
+NTAPI
+InPortUnload(
+    _In_ PDRIVER_OBJECT DriverObject)
+{
+    PAGED_CODE();
+
+    DPRINT("%s(%p)\n", __FUNCTION__, DriverObject);
+
+    RtlFreeUnicodeString(&DriverRegistryPath);
+}
+
+NTSTATUS
+NTAPI
+InPortPower(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _Inout_ PIRP Irp)
+{
+    NTSTATUS Status;
+    PINPORT_DEVICE_EXTENSION DeviceExtension = DeviceObject->DeviceExtension;
+
+    DPRINT("%s(%p, %p) %X\n", __FUNCTION__, DeviceObject, Irp,
+           IoGetCurrentIrpStackLocation(Irp)->MinorFunction);
+
+    PoStartNextPowerIrp(Irp);
```

Comment:
Not calling this _immediately_ before `IoCompleteRequest`/`IoSkipCurrentIrpStackLocation` seems odd, but I can't give a concrete reason why it would be. The documentation doesn't seem to specify this in much detail.